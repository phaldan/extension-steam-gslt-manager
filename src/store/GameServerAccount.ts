import { action, observable } from 'mobx';

export default class GameServerAccount {
  public steamid: string;
  public appid: number;

  @observable
  public expired: boolean = false;

  @observable
  public lastLogon?: Date;

  @observable
  public memo: string;

  @observable
  public token: string;

  @observable
  public locked: boolean = false;

  constructor({ steamid, appid, token, memo, lastLogon, expired }) {
    this.steamid = steamid;
    this.appid = appid;
    this.token = token;
    this.memo = memo;
    this.lastLogon = lastLogon ? new Date(lastLogon) : undefined;
    this.expired = expired;
  }

  @action
  public toggleLock(locked: boolean) {
    this.locked = locked;
  }

  @action
  public updateFromJson({ appid, token, memo, lastLogon, expired }) {
    this.appid = appid;
    this.token = token;
    this.memo = memo;
    this.lastLogon = lastLogon;
    this.expired = expired;
  }
}
