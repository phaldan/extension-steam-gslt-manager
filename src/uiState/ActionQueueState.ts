import { action, computed, observable } from 'mobx';
import ActionQueue from './ActionQueue';

export default class ActionQueueState {
  private removeDelay = 3000;

  @observable
  private queues: ActionQueue[] = [];

  @action
  public create(description: string, data: string[]): ActionQueue {
    const entry = new ActionQueue(this, this.removeDelay, description, data);
    this.queues.push(entry);
    return entry;
  }

  @computed
  public get running(): ActionQueue[] {
    return this.queues;
  }

  @action
  public async remove(id: string) {
    this.queues = this.queues.filter((q) => q.id !== id);
  }

  public setRemoveDelay(removeDelay: number): void {
    this.removeDelay = removeDelay;
  }
}
