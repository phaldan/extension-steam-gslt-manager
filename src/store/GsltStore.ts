import GameServerAccount from './GameServerAccount';

export default interface GsltStore {

  /**
   * Return current list of GameServerAccount.
   */
  readonly tokenAccounts: GameServerAccount[];

  readonly isLoggedIn: boolean;

  readonly isInitialized: boolean;

  /**
   * Load token accounts from steam.
   */
  loadAccounts(): Promise<void>;

  /**
   * Remove GSLT account.
   * @param account GameServerAccount instance
   */
  removeAccount(account: GameServerAccount): Promise<GameServerAccount>;

  /**
   * Remove GSLT accounts.
   * @param accounts List of GameServerAccount instances
   */
  removeAccounts(accounts: GameServerAccount[]): Array<Promise<GameServerAccount>>;

  /**
   * Create a new token for GSLT account.
   * @param account GameServerAccount instance
   */
  regenerateToken(account: GameServerAccount): Promise<GameServerAccount>;

  /**
   * Create a new token for GSLT accounts.
   * @param accounts List of GameServerAccount instances.
   */
  regenerateTokens(accounts: GameServerAccount[]): Array<Promise<GameServerAccount>>;

  /**
   * Change memo of GSLT account.
   * @param account GameServerAccount instance
   * @param memo Description for a GSLT
   */
  updateMemo(account: GameServerAccount, memo: string): Promise<GameServerAccount>;

  /**
   * Create multiple GSLT accounts.
   * @param amount Number of GSLT to create
   * @param appid Steam AppID
   * @param memo Description for a GSLT
   */
  createAccounts(amount: number, appid: string, memo: string): Array<Promise<void>>;
}
