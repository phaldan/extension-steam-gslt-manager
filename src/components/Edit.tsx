import { observer } from 'mobx-react';
import * as React from 'react';
import { FormControl } from 'react-bootstrap';
import GameServerAccount from '../store/GameServerAccount';
import GsltStore from '../store/GsltStore';
import ActionQueueState from '../uiState/ActionQueueState';
import ActionModal from './ActionModal';

interface EditProps {
  store: GsltStore;
  account: GameServerAccount;
  disabled: boolean;
  actions: ActionQueueState;
}

interface EditState {
  value: string;
}

@observer
export default class Edit extends React.Component<EditProps, EditState> {
  constructor(props) {
    super(props);
    this.state = { value: this.props.account.memo };
  }

  public render() {
    const { account, disabled } = this.props;
    return (
      <ActionModal
        buttonStyle="primary"
        buttonSize="small"
        buttonIcon="edit"
        title="GSLT Memo"
        action="Update"
        onAction={this.save}
        disabled={disabled}
      >
        <FormControl
          componentClass="textarea"
          className="js-memo"
          defaultValue={account.memo}
          rows="10"
          maxLength="256"
          onChange={this.updateMemo}
        />
      </ActionModal>
    );
  }

  private save = async (): Promise<void> => {
    const { value } = this.state;
    const { store, account, actions } = this.props;
    const queue = actions.create('Update memo:', [account.steamid]);
    await store.updateMemo(account, value);
    queue.progress(account.steamid);
  }

  private updateMemo = (event): void => {
    this.setState({ value: event.target.value });
  }
}
