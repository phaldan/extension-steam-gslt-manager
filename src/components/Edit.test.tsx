import { mount, shallow } from 'enzyme';
import * as React from 'react';
import GameServerAccount from '../store/GameServerAccount';
import GsltStore from '../store/GsltStore';
import ActionQueueState from '../uiState/ActionQueueState';
import delay from '../utils/delay';
import Edit from './Edit';

class GsltStoreStub implements GsltStore {
  public updateReturn: Promise<GameServerAccount>;
  public updateMock = jest.fn();

  loadAccounts(): Promise<void> {
    throw new Error("Method not implemented.");
  }
  get tokenAccounts(): GameServerAccount[] {
    throw new Error("Method not implemented.");
  }
  removeAccount(account: GameServerAccount): Promise<GameServerAccount> {
    throw new Error("Method not implemented.");
  }
  removeAccounts(accounts: GameServerAccount[]): Promise<GameServerAccount>[] {
    throw new Error("Method not implemented.");
  }
  regenerateToken(account: GameServerAccount): Promise<GameServerAccount> {
    throw new Error("Method not implemented.");
  }
  regenerateTokens(accounts: GameServerAccount[]): Promise<GameServerAccount>[] {
    throw new Error("Method not implemented.");
  }
  updateMemo(account: GameServerAccount, memo: string): Promise<GameServerAccount> {
    this.updateMock(account, memo);
    return this.updateReturn;
  }
  createAccounts(amount: number, appid: string, memo: string): Promise<void>[] {
    throw new Error("Method not implemented.");
  }
  get isLoggedIn(): boolean {
    throw new Error("Method not implemented.");
  }
  get isInitialized(): boolean {
    throw new Error("Method not implemented.");
  }
}

describe('<Edit>', () => {
  let account: GameServerAccount;
  let store: GsltStoreStub;
  let queue: ActionQueueState;

  function createComponent(account, state, disabled) {
    return shallow(<Edit account={account} store={state} actions={queue} disabled={disabled} />);
  }

  function expectEdit(target, modalIsDisabled) {
    expect(target.find('ActionModal').props().disabled).toEqual(modalIsDisabled);
    expect(target.find('.js-memo').props().defaultValue).toEqual('Awesome');
    expect(account.memo).toEqual('Awesome');
    expect(queue.running).toHaveLength(0);
    expect(target).toMatchSnapshot();
  }

  beforeEach(() => {
    account = new GameServerAccount({
      appid: 730,
      expired: false,
      lastLogon: undefined,
      memo: 'Awesome',
      steamid: '',
      token: '',
    });
    store = new GsltStoreStub();
    queue = new ActionQueueState();
    queue.setRemoveDelay(0);
  });

  test('disabled ActionModal', () =>  {
    const target = createComponent(account, store, true);

    expectEdit(target, true);
    expect(store.updateMock).not.toHaveBeenCalled();
  });

  describe('enabled ActionModal', () => {
    let target;
    beforeEach(() => {
      target = createComponent(account, store, false);
    });

    test('save memo', async () =>  {
      target.find('ActionModal').simulate('action');

      await delay(10);
      expectEdit(target, false);
      expect(store.updateMock).toHaveBeenCalledWith(account, 'Awesome');
    });

    test('change and save memo', async () =>  {
      target.find('.js-memo').simulate('change', {target: {value: 'Changed'}});
      target.find('ActionModal').simulate('action');

      await delay(10);
      expectEdit(target, false);
      expect(store.updateMock).toHaveBeenCalledWith(account, 'Changed');
    });
  });
});
