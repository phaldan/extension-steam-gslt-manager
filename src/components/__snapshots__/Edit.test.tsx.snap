// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Edit> disabled ActionModal 1`] = `
<ActionModal
  action="Update"
  buttonIcon="edit"
  buttonSize="small"
  buttonStyle="primary"
  close="Abort"
  disabled={true}
  onAction={[Function]}
  title="GSLT Memo"
>
  <FormControl
    bsClass="form-control"
    className="js-memo"
    componentClass="textarea"
    defaultValue="Awesome"
    maxLength="256"
    onChange={[Function]}
    rows="10"
  />
</ActionModal>
`;

exports[`<Edit> enabled ActionModal change and save memo 1`] = `
<ActionModal
  action="Update"
  buttonIcon="edit"
  buttonSize="small"
  buttonStyle="primary"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  title="GSLT Memo"
>
  <FormControl
    bsClass="form-control"
    className="js-memo"
    componentClass="textarea"
    defaultValue="Awesome"
    maxLength="256"
    onChange={[Function]}
    rows="10"
  />
</ActionModal>
`;

exports[`<Edit> enabled ActionModal save memo 1`] = `
<ActionModal
  action="Update"
  buttonIcon="edit"
  buttonSize="small"
  buttonStyle="primary"
  close="Abort"
  disabled={false}
  onAction={[Function]}
  title="GSLT Memo"
>
  <FormControl
    bsClass="form-control"
    className="js-memo"
    componentClass="textarea"
    defaultValue="Awesome"
    maxLength="256"
    onChange={[Function]}
    rows="10"
  />
</ActionModal>
`;
